cache/snapshot.go

```diff
@@ -0,0 +1,163 @@
+package cache
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+
+	"github.com/samsarahq/taskrunner/shell"
+)
+
+type snapshotter struct {
+	CommitFunc           func() (sha string, err error)
+	DiffFunc             func(string) (diffFiles []string, err error)
+	UncommittedFilesFunc func() (newFiles []string, modifiedFiles []string, err error)
+	HashFunc             func(string) (hash string, err error)
+}
+
+func newSnapshotter(ctx context.Context, shellRun shell.ShellRun) *snapshotter {
+	client := gitClient{ctx: ctx, shellRun: shellRun}
+
+	return &snapshotter{
+		DiffFunc:             client.diff,
+		CommitFunc:           client.currentCommit,
+		UncommittedFilesFunc: client.uncomittedFiles,
+		HashFunc:             hashSum,
+	}
+}
+
+// compare compares against another snapshot with the current state.
+func (c *snapshotter) diff(previous *snapshot) ([]string, error) {
+	current, err := c.snapshot(false)
+	if err != nil {
+		return nil, err
+	}
+
+	diffFiles, err := c.DiffFunc(previous.CommitSha)
+	if err != nil {
+		return nil, err
+	}
+
+	var dirtyFiles []string
+
+	// Only mark committed files as dirty (uncommitted files are handled below).
+	for _, f := range diffFiles {
+		if previous.shaFor(f) == "" {
+			dirtyFiles = append(dirtyFiles, f)
+		}
+	}
+
+	// Because we diff against the last commit, any files that were not committed the
+	// last time a snapshot was recorded needs to compare hashes instead.
+	for _, file := range previous.UncommittedFiles {
+		// If the file isn't currently uncommitted, rehash.
+		md5 := current.shaFor(file.Path)
+		if md5 == "" {
+			md5, _ = hashSum(file.Path)
+		}
+		if md5 == "" || md5 != file.MD5 {
+			dirtyFiles = append(diffFiles, file.Path)
+		}
+	}
+
+	for _, file := range current.UncommittedFiles {
+		// Any uncommitted file that either has a different MD5 hash or wasn't recorded in the
+		// previous snapshot counts as different.
+		if sha := previous.shaFor(file.Path); sha == "" || sha != file.MD5 {
+			dirtyFiles = append(diffFiles, file.Path)
+		}
+	}
+
+	return dirtyFiles, err
+}
+
+// snapshot takes a snapshot of the current state.
+func (c *snapshotter) snapshot(withChanged bool) (*snapshot, error) {
+	var err error
+	s := snapshot{}
+
+	s.CommitSha, err = c.CommitFunc()
+	if err != nil {
+		return nil, err
+	}
+
+	newFiles, modifiedFiles, err := c.UncommittedFilesFunc()
+	if err != nil {
+		return nil, err
+	}
+
+	// We only want modified files for recording snapshots, not for diffing against an older
+	// snapshot (since we can rely on git diff to handle those).
```

Comment:
could you give an example to help explain this comment?

Reply:
I'll think about a useful example there